Rewriting SS Model as a Set of Functions

anu1217 opened the pull request:
Adding OpenMC scripts with functions that create .xml files

anu1217 committed:
Add files via upload

Adding OpenMC scripts with functions that create .xml files

anu1217 committed:
Update and rename OpenMC_xml.py to OpenMCSphericalShellFunctions/OpenMC_xml.py

Adding additional folder to store scripts

gonuke commented on OpenMCSphericalShellFunctions/OpenMC_xml.py:
```diff
@@ -0,0 +1,123 @@
+import openmc
+import argparse
+
+parser = argparse.ArgumentParser(description="Specify required inputs: file path to ALARA Element Library, element name, inner radius [cm], outer_radius [cm], particle energy [eV]")
+
+#Required user inputs:
+parser.add_argument('--filepath', type=str, required=True)    
+parser.add_argument('--element', type=str, required=True)
+parser.add_argument('--inner_radius', type=float, required=True)
+parser.add_argument('--outer_radius', type=float, required=True)
+parser.add_argument('--energy', type=float, required=True)
+
+args = parser.parse_args()
```
Useful to put this in a method of it's own:
```suggestion
def define_parser():
    parser = argparse.ArgumentParser(description="Specify required inputs: file path to ALARA Element Library, element name, inner radius [cm], outer_radius [cm], particle energy [eV]")

    #Required user inputs:
    parser.add_argument('--filepath', type=str, required=True)    
    parser.add_argument('--element', type=str, required=True)
    parser.add_argument('--inner_radius', type=float, required=True)
    parser.add_argument('--outer_radius', type=float, required=True)
    parser.add_argument('--energy', type=float, required=True)

    return parser.parse_args()

```

gonuke commented on OpenMCSphericalShellFunctions/OpenMC_xml.py:
```diff
@@ -0,0 +1,123 @@
+parser = argparse.ArgumentParser(description="Specify required inputs: file path to ALARA Element Library, element name, inner radius [cm], outer_radius [cm], particle energy [eV]")
+
+#Required user inputs:
+parser.add_argument('--filepath', type=str, required=True)    
```
This appears to the path to the ALARA element library?  If so, it should have a better name than just `filepath`

gonuke commented on OpenMCSphericalShellFunctions/OpenMC_xml.py:
```diff
@@ -0,0 +1,123 @@
+import openmc
+import argparse
+
+parser = argparse.ArgumentParser(description="Specify required inputs: file path to ALARA Element Library, element name, inner radius [cm], outer_radius [cm], particle energy [eV]")
```
PEP8 - use the "black" autoformatter

gonuke commented on OpenMCSphericalShellFunctions/OpenMC_xml.py:
```diff
@@ -0,0 +1,123 @@
+
+# Create materials & export to XML:
+#Simulating tungsten shell:
+print(E_1)   
```
Probably don't want print statements in our final scripts

gonuke commented on OpenMCSphericalShellFunctions/OpenMC_xml.py:
```diff
@@ -0,0 +1,123 @@
+
+args = parser.parse_args()
+
+fp = args.filepath
```
see below...

gonuke commented on OpenMCSphericalShellFunctions/OpenMC_xml.py:
```diff
@@ -0,0 +1,123 @@
+    #return OpenMC_Materials, OpenMC_Geometry, OpenMC_Energy, OpenMC_Settings, OpenMC_Tallies, OpenMC_Universe
+    return OpenMC_Materials, *OpenMC_Geometry, *OpenMC_Energy, OpenMC_Settings, *OpenMC_Tallies
+
+M_1, geometry, Void, Shell, Cells, PointSource, Prob, sets, tall, neutron_tally, spectrum_tally = export_to_xml(E_1, R_1, R_2, eg)
```
I'm not sure what you do with all of these returned things, but also suggest something like this.... 
```suggestion
if __name__ == "__main__":
    args = define_parser()
    fp = args.filepath
    E_1 = args.element
    R_1 = args.inner_radius
    R_2 = args.outer_radius
    eg = args.energy

    with open(""+fp+"") as ALARA_Lib:
        Lib_Lines = ALARA_Lib.readlines()
    
    M_1, geometry, Void, Shell, Cells, PointSource, Prob, sets, tall, neutron_tally, spectrum_tally = export_to_xml(E_1, R_1, R_2, eg)

```

gonuke reviewed (changes requested):
A few comments here to get started

anu1217 committed:
Update OpenMCSphericalShellFunctions/OpenMC_xml.py

Add separate method for parser

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Remove print statement

anu1217 committed:
Delete OpenMCSphericalShellFunctions/OpenMC_xml.py

Replacing with individual files that make OpenMC Materials, Geometry, and Model

anu1217 committed:
Merge branch 'cnerg:main' into xml_split

anu1217 committed:
Add files via upload

anu1217 committed:
Update and rename OpenMC_SS_Geometry.py to SphericalShell/Split_into_Functions/OpenMC_Model/OpenMC_SS_Geometry.py

anu1217 committed:
Update and rename OpenMC_SS_YAML.yaml to SphericalShell/Split_into_Functions/OpenMC_Model/OpenMC_SS_YAML.yaml

anu1217 committed:
Update and rename OpenMC_SS_Material.py to SphericalShell/Split_into_Functions/OpenMC_Model/OpenMC_SS_Material.py

anu1217 committed:
Update and rename OpenMC_Source_Tallies_Model.py to SphericalShell/Split_into_Functions/OpenMC_Model/OpenMC_Source_Tallies_Model.py

anu1217 commented:
I've broken up the existing Spherical Shell script into a Materials, Geometry, and a Source/Tallies/Model. The depletion and post-processing will come in subsequent PRs

gonuke commented:
I think I anticipated that this would replace the version in `SphericalShell/OpenMC_Input`, rather than keeping two versions of the same thing.  This new code implements best practices so should be a reasonable replacement of the old code.

anu1217 commented:
I'll add the depletion and post-processing scripts here as well since SS_Activation_OpenMC is getting replaced

anu1217 committed:
Add files via upload

anu1217 committed:
Update and rename SS_Depletion.py to SphericalShell/Split_into_Functions/OpenMC_Depletion/SS_Depletion.py

anu1217 committed:
Update and rename SphericalShell/Split_into_Functions/OpenMC_Model/OpenMC_SS_YAML.yaml to SphericalShell/Split_into_Functions/OpenMC_SS_YAML.yaml

Move YAML to directory above model & depletion

anu1217 committed:
Update OpenMC_SS_YAML.yaml with depletion parameters

anu1217 committed:
Delete SphericalShell/OpenMC_Input/SS_Activation_OpenMC.py

This is being replaced by a set of xml files that make the model, perform depletion, and analyze results

anu1217 committed:
Update and rename SphericalShell/Split_into_Functions/OpenMC_Depletion/SS_Depletion.py to SphericalShell/OpenMC_Input/Depletion

anu1217 committed:
Update and rename SphericalShell/Split_into_Functions/OpenMC_Model/OpenMC_SS_Geometry.py to SphericalShell/OpenMC_Input/Model

anu1217 committed:
Update and rename SphericalShell/OpenMC_Input/Depletion to SphericalShell/OpenMC_Input/OpenMC_Depletion/SS_Depletion

anu1217 committed:
Update and rename SphericalShell/OpenMC_Input/Model to SphericalShell/OpenMC_Input/OpenMC_Model/SS_Geometry

anu1217 committed:
Update and rename SphericalShell/Split_into_Functions/OpenMC_Model/OpenMC_SS_Material.py to SphericalShell/OpenMC_Input/SS_Material.py

anu1217 committed:
Update and rename SS_Depletion to SS_Depletion.py

anu1217 committed:
Update and rename SS_Geometry to SS_Geometry.py

anu1217 committed:
Update and rename SphericalShell/OpenMC_Input/SS_Material.py to SphericalShell/OpenMC_Input/OpenMC_Model/SS_Material.py

anu1217 committed:
Update and rename OpenMC_Source_Tallies_Model.py to SS_Source_Tallies_Model.py

anu1217 committed:
Update and rename OpenMC_SS_YAML.yaml to OpenMC_SS_YAML.yaml

anu1217 committed:
Delete SphericalShell/OpenMC_Input/SphericalShell.py

gonuke commented on SphericalShell/OpenMC_Input/OpenMC_Model/SS_Source_Tallies_Model.py:
```diff
@@ -0,0 +1,41 @@
+import yaml
+
+file = open("OpenMC_SS_YAML.yaml")
+data = yaml.safe_load(file)
```
you don't use this data in the end

anu1217 replied:
Accidentally left this in - it's been deleted now. I'm assuming the user will import the file when it's time to call these functions

gonuke reviewed (changes requested):
This looks pretty good now, but you need a script that pulls it all together by importing and using all these scripts

anu1217 committed:
Delete reference to YAML file

anu1217 committed:
Update inputs and outputs to SS_Material.py

anu1217 committed:
Update SS_Source_Tallies_Model.py - assign particle filter to tallies

anu1217 committed:
Update OpenMC_SS_YAML.yaml

anu1217 committed:
Remove material volume assignment

anu1217 committed:
Update SS_Material.py

anu1217 committed:
Add files via upload

- SS_Model_to_xml calls all functions required to make and export a model xml
- SS_Tally_PostProcessing & SS_Depletion_PostProcessing use functions to perform post-processing
- SS_Post_Processing_YAML contains inputs required for post-processing
- SS_all_PostProcessing calls all functions from the Tally and Depletion post-processing scripts
- SS_Run_Depletion calls the depletion simulation function from SS_Depletion

anu1217 committed:
Update and rename SS_Depletion_PostProcessing.py to SphericalShell/OpenMC_Output_Processing/SS_Depletion_PostProcessing.py

anu1217 committed:
Update and rename SS_Model_to_xml.py to SphericalShell/OpenMC_Input/SS_Model_to_xml.py

anu1217 committed:
Update and rename SS_Post_Processing_YAML.yaml to SphericalShell/OpenMC_Output_Processing/SS_Post_Processing_YAML.yaml

anu1217 committed:
Update and rename SS_Run_Depletion.py to SphericalShell/OpenMC_Input/OpenMC_Model/SS_Run_Depletion.py

anu1217 committed:
Update and rename SS_Run_Depletion.py to SS_Run_Depletion.py

anu1217 committed:
Update and rename SS_Tally_PostProcessing.py to SphericalShell/OpenMC_Output_Processing/SS_Tally_PostProcessing.py

anu1217 committed:
Update and rename SS_all_PostProcessing.py to SphericalShell/OpenMC_Output_Processing/SS_all_PostProcessing.py

gonuke commented on SphericalShell/OpenMC_Input/SS_Model_to_xml.py:
```diff
@@ -0,0 +1,42 @@
+num_particles = inputs['settings_info']['num_particles']
+run_mode = inputs['settings_info']['run_mode']
+
+def run_ss_scripts(elelib_fp, element, thickness, inner_radius, energy, total_batches, inactive_batches, num_particles, run_mode) :
```
Pass in the full settings dictionary
```suggestion
def run_ss_scripts(elelib_fp, element, thickness, inner_radius, energy, settings) :
```

gonuke commented on SphericalShell/OpenMC_Input/SS_Model_to_xml.py:
```diff
@@ -0,0 +1,42 @@
+    mss = make_spherical_shell(element[0][0], thickness, inner_radius)
+    #OpenMC_Source_Tallies_Model :
+    ms = make_source(energy)
+    sets = settings(ms[0], ms[1], total_batches, inactive_batches, num_particles, run_mode)
```
Use the data from the settings dictionary
```suggestion
    sets = settings(ms[0], ms[1], settings['total_batches'], 
          settings['inactive_batches'], 
          settings['num_particles'], 
          settings['run_mode'])
```

gonuke commented on SphericalShell/OpenMC_Input/SS_Model_to_xml.py:
```diff
@@ -0,0 +1,42 @@
+from OpenMC_SS_Geometry import make_spherical_shell
+from OpenMC_Source_Tallies_Model import make_source, settings, tallies, create_openmc_model
+
+file = open("OpenMC_SS_YAML.yaml")
+inputs = yaml.safe_load(file)
```
```suggestion
with open("OpenMC_SS_YAML.yaml") as file:
    inputs = yaml.safe_load(file)
```

gonuke reviewed (changes requested):
we discussed some redesign for the SS_Model_to_xml.py script.

anu1217 committed:
Update SphericalShell/OpenMC_Input/SS_Model_to_xml.py

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Update SphericalShell/OpenMC_Input/SS_Model_to_xml.py

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Compile helper functions using main function

anu1217 committed:
Change function names to make_settings and make_tallies

anu1217 committed:
Compile helper functions using main function

anu1217 committed:
Compile post-processing scripts into main()

anu1217 committed:
Delete SphericalShell/OpenMC_Input/OpenMC_Depletion/SS_Run_Depletion.py

anu1217 committed:
Update SS_Tally_PostProcessing.py

anu1217 committed:
Update SS_Depletion_PostProcessing.py

anu1217 committed:
Initiate depletion from main() function

anu1217 committed:
Remove commented line

anu1217 committed:
Add model.xml as input

anu1217 committed:
Remove returned variables

gonuke commented on SphericalShell/OpenMC_Input/OpenMC_Model/SS_Material.py:
```diff
@@ -0,0 +1,40 @@
+    '''
+    inputs:
+        element: elemental symbol of chosen element (str)
+        density_dict: dictionary with key = elemental symbol & value = density [g/cm^3]
```
Your docstring doesn't include the geometry info.... but I wonder if that should be done elsewhere.  I am guessing this is only necessary for the depletion/activation calculation.  Should we do it in a different way to keep the separation of concerns of just making a new material?

anu1217 replied:
The volume is used for depletion and during post-processing, where flux data is normalized by volume. Calculating the volume separately for both of these might be best for separation of concerns...I think I was trying to avoid having to import the thickness/radius on each occasion

gonuke commented on SphericalShell/OpenMC_Input/SS_Model_to_xml.py:
```diff
@@ -0,0 +1,41 @@
+    settings_info = inputs['settings_info']
+    # for OpenMC_SS_Material
+    aed = alara_element_densities(inputs['elelib_fp'])
+    element = make_element(inputs['element'], 
```
While this method only creates a single material from a single element, it may be more clear to name this variable `materials` or something similar since it is an OpenMC `materials` object

gonuke replied:
If you thought it would offer some clarity, you could also have:
```
materials, volume = make_element(....)
element = materials[0]
```

gonuke commented on SphericalShell/OpenMC_Input/SS_Model_to_xml.py:
```diff
@@ -0,0 +1,41 @@
+                           geometry['inner_radius'], 
+                           geometry['thickness'])
+    # for OpenMC_SS_Geometry
+    mss = make_spherical_shell(element[0][0], 
```
Is the first `[0]` because your `make_element()` method also returns the volume?  If so, you may want to capture those when they are returned: `element, volume = make_element(....)` or similar

gonuke replied:
Also probably best to just call this `spherical_shell_geom` or something

gonuke commented on SphericalShell/OpenMC_Input/SS_Model_to_xml.py:
```diff
@@ -0,0 +1,41 @@
+                               geometry['thickness'], 
+                               geometry['inner_radius'])
+    # for OpenMC_Source_Tallies_Model
+    ms = make_source(inputs['particle_energy'])
```
For readability, I recommend a different variable name
```suggestion
    source_dists = make_source(inputs['particle_energy'])
```

gonuke commented on SphericalShell/OpenMC_Input/OpenMC_Model/SS_Source_Tallies_Model.py:
```diff
@@ -0,0 +1,35 @@
+import openmc
+    
+def make_source(energy):
```
I wonder if this should be named `make_point_source`?  It also doesn't actually make a suorce, but just makes two source distributions.  So maybe `make_point_source_dists`?

gonuke replied:
Alternatively, you could actually make/return an `openmc.Source()` here and return that instead.

gonuke commented on SphericalShell/OpenMC_Input/OpenMC_Model/SS_Source_Tallies_Model.py:
```diff
@@ -0,0 +1,35 @@
+    spectrum_tally.scores = ['flux']
+    
+    talls = openmc.Tallies([neutron_tally, spectrum_tally])
+    return talls
```
```suggestion
    return openmc.Tallies([neutron_tally, spectrum_tally])
```

gonuke commented on SphericalShell/OpenMC_Input/OpenMC_Model/SS_Source_Tallies_Model.py:
```diff
@@ -0,0 +1,35 @@
+    talls = openmc.Tallies([neutron_tally, spectrum_tally])
+    return talls
+
+def create_openmc_model(mats_object, geom_object, tallies_object, settings_object):
```
Not sure this needs its own method since it does nothing other than what the openmc method already does

anu1217 replied:
Would it be better to make this model object in SS_Model_to_xml instead?

anu1217 replied:
The Model object is now being made in SS_Model_to_xml

gonuke commented on SphericalShell/OpenMC_Input/SS_Model_to_xml.py:
```diff
@@ -0,0 +1,41 @@
+    # tallied cells = all cells with non-void material
+    tallied_cells = list(mss.get_all_material_cells().values())
+    talls = make_tallies(tallied_cells)
+    com = create_openmc_model(element[0], mss, talls, sets)
```
I recommend just calling this `model` (and probably just calling the openmc function directly - see related comment)
```suggestion
    model = create_openmc_model(element[0], mss, talls, sets)
```

gonuke commented on SphericalShell/OpenMC_Input/SS_Model_to_xml.py:
```diff
@@ -0,0 +1,41 @@
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--yaml_filepath', required=True, help="Path to YAML file containing required inputs (str)")
+    args = parser.parse_args()
```
Its common to put these in a different function for separation of concerns.  By separating file input in its different forms from action it helps make testing a little simpler.

gonuke commented on SphericalShell/OpenMC_Input/SS_Model_to_xml.py:
```diff
@@ -0,0 +1,41 @@
+    args = parser.parse_args()
+    yaml_filepath = args.yaml_filepath
+    with open(yaml_filepath, 'r') as file:
+        inputs = yaml.safe_load(file)
```
These lines might also go in a separate method for separation of concerns.

gonuke commented on SphericalShell/OpenMC_Input/SS_Model_to_xml.py:
```diff
@@ -0,0 +1,41 @@
+from OpenMC_SS_Geometry import make_spherical_shell
+from OpenMC_Source_Tallies_Model import make_source, settings, tallies, create_openmc_model
+
+def main():
```
For separation of concerns, you might think of a main function that was:
```
def main():
    args = parse_args() # this is a short method comprised of your first 3 lines

    input_data = read_yaml(args[yaml_filepath])  # this is a short method comprised of the next 3 lines

    model = build_model(input_data)  # this is a method that holds most of your code

    model.export_to_model_xml()
```
Although testing may not seem critical, this means that you can easily test that `build_model()` does what you want with a given `input_data` dictionary without having to make a YAML input file for each variation.

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_Tally_PostProcessing.py:
```diff
@@ -0,0 +1,56 @@
+        flux_tally_id : id of OpenMC Tally object that scores flux with energy filter
+        energy_filter_index : index of EnergyFilter within the list of applied filters
+    '''
+    flux_tally = tallies[flux_tally_id]
```
Since this is the first thing you do, you could consider just passing in the singly tally of interest rather than all the tallies and the ID

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_Tally_PostProcessing.py:
```diff
@@ -0,0 +1,56 @@
+    flux_tally = tallies[flux_tally_id]
+    flux_tally_values = flux_tally.get_values(value = 'mean').ravel()
+    energy_filter = flux_tally.filters[energy_filter_index]
+    energy_bins = energy_filter.bins[:, 0]
```
```suggestion
    energy_bins = flux_tally.filters[energy_filter_index].bins[:, 0]
```

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_Tally_PostProcessing.py:
```diff
@@ -0,0 +1,56 @@
+        tally_array = np.array(tally_array)   
+    return tallies, tally_array
+    
+def plot_flux_tally(tallies, flux_tally_id, energy_filter_index) :
```
```suggestion
def plot_flux_spectrum(tallies, flux_tally_id, energy_filter_index) :
```

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_all_PostProcessing.py:
```diff
@@ -0,0 +1,47 @@
+    yaml_model_path = args.yaml_model_path
+    
+    with open(yaml_postprocess_path, 'r') as yaml_pp :
+        pp_inputs = yaml.safe_load(yaml_pp)
```
See comments elsewhere about separation of concerns and `main` being a method that calls a variety of other methods

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_Tally_PostProcessing.py:
```diff
@@ -0,0 +1,56 @@
+        for tally_id, tally in tallies.items():
+            tally_values = tally.get_values(value="mean").ravel()
+            tally_array.append(np.array(tally_values))
+        tally_array = np.array(tally_array)   
```
These lines don't need to be part of the `with` block

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_Tally_PostProcessing.py:
```diff
@@ -0,0 +1,56 @@
+    '''
+    with openmc.StatePoint(statepoint_file_path) as sp:
+        tallies = sp.tallies
+        tally_array = []
+        for tally_id, tally in tallies.items():
+            tally_values = tally.get_values(value="mean").ravel()
+            tally_array.append(np.array(tally_values))
+        tally_array = np.array(tally_array)   
```
Aren't the different tallies in this list different dimensions from each other? (e.g. total reaction rates and flux tallies vs spectrum)  What's the intent in making it a numpy array?

anu1217 replied:
I was trying to avoid having data stored in lists in general, but it doesn't make a difference as far as processing is concerned

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_all_PostProcessing.py:
```diff
@@ -0,0 +1,47 @@
+    thickness = model_inputs['geom_info']['thickness'] 
+    
+    # Tally post-processing:
+    etv = extract_tally_values(filepaths['statepoint_file_path'])
```
If you return multiple values with different intents, it's best to separate them ASAP.  One way that should work is:
```suggestion
    tallies, tally_array = extract_tally_values(filepaths['statepoint_file_path'])
```

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_all_PostProcessing.py:
```diff
@@ -0,0 +1,47 @@
+                          indices['flux_tally_id'], 
+                          indices['energy_filter_index']) 
+    aed = alara_element_densities(elelib_fp)  
+    me = make_element(element, aed, inner_radius, thickness)
+    mat_vol = me[1]
```
Probably a lot of work to just get the volume?  Maybe calculating this volume should be it's own method? [separation of concerns]

anu1217 replied:
The volume is calculated separately in the depletion and tally post-processing scripts now

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_all_PostProcessing.py:
```diff
@@ -0,0 +1,47 @@
+    mat_vol = me[1]
+    sd = save_data(pft[0], mat_vol, etv[1])
+    
+    # Depletion post-processing:
+    en = extract_nuclides(filepaths['dep_file_path'], 
+                          units['time_units'], 
+                          indices['depletable_mat_index'])    
+    psd = plot_save_data(en[0], en[1], en[2], en[3], units['nuc_units'])
```
We should think about how to separate transport post-processing from depletion post-processing

anu1217 replied:
Transport and depletion post-processing are now part of main() but belong to separate inner functions

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_Depletion_PostProcessing.py:
```diff
@@ -0,0 +1,62 @@
+    nuclide_array = [nuc for nuc in nuclide_array if nuc is not None]
+    return nuclide_array, materials_object, dep_results, time_steps
+
+def plot_save_data(nuclide_array, materials_object, dep_results, time_steps, nuc_units) :
```
Separation of concerns: making a plot and saving data should be separate methods, and maybe a method to extract that data from the `dep_results` into a convenient data structure to pass to each of these methods

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_Depletion_PostProcessing.py:
```diff
@@ -0,0 +1,62 @@
+    for step in range(len(time_steps)):
+        # Obtain a materials object with depletion information at each timestep
+        materials_object = dep_results.export_to_materials(step)[depletable_mat_index]
+        all_nuc = materials_object.get_nuclides()      
+        for nuclide in all_nuc :   
```
```suggestion
        for nuclide in materials_object.get_nuclides()  :   
```

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_Depletion_PostProcessing.py:
```diff
@@ -0,0 +1,62 @@
+                nuclide_array = np.append(nuclide_array, nuclide) 
+                       
+    # Remove None element from array initialization    
+    nuclide_array = [nuc for nuc in nuclide_array if nuc is not None]
```
How do we end up with `None` in this list/array?

anu1217 replied:
This comes from initialization of the associated np.ndarray

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_Depletion_PostProcessing.py:
```diff
@@ -0,0 +1,62 @@
+                       
+    # Remove None element from array initialization    
+    nuclide_array = [nuc for nuc in nuclide_array if nuc is not None]
+    return nuclide_array, materials_object, dep_results, time_steps
```
This only returns the `materials_object` from the last time step, is that intentional/necessary?

anu1217 replied:
From what I understand, it doesn't matter which time step `materials_object` is from, as it is only used in conjunction with `get_atoms()`, which gives nuclide concentration vs. time based on the material the specified nuclide came from. I see the same results if I specify a material id for `get_atoms` instead of a materials object.

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_Depletion_PostProcessing.py:
```diff
@@ -0,0 +1,62 @@
+import numpy as np
+import matplotlib.pyplot as plt
+
+def extract_nuclides(dep_file_path, time_units, depletable_mat_index) : 
```
It seems like the purpose of this is to collect a list of all the nuclides that appear, excluding the list of stable nuclides at beginning of life.  Is that right?  Do don't seem to actually record/return any data about those nuclides as a function of time.

anu1217 replied:
This function just stores all the new nuclides that were created (and any unstable ones at the beginning of operation), so that they can be easily retrieved later. The get_atoms method (used in plot_save_data) returns nuclide concentration as a function of time.

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_Depletion_PostProcessing.py:
```diff
@@ -0,0 +1,62 @@
+                half_life = openmc.data.half_life(nuclide)
+                if half_life == None:
+                    stable_init_nuc = np.append(stable_init_nuc, nuclide)
+            if nuclide not in nuclide_array and nuclide not in stable_init_nuc:        
```
You might consider using a `set()` for this rather than a numpy array.  With sets I would propose that you simply build a set of all nuclides - you don't need to check wats already in the set because sets do that automatically. You can also build a set of all the stable nuclides at the initial time.  Then, at the end, you can remove the set of stables from the set of all nuclides.  The code will probably be much simpler and straightforward.

gonuke reviewed (changes requested):
This gets better/cleaner with each step. Plugging away at more readability,  streamlining and separation of concerns

anu1217 committed:
Update SphericalShell/OpenMC_Output_Processing/SS_Depletion_PostProcessing.py

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Change nuclide_array to nuclide_set

anu1217 committed:
Rename make_element as make_materials

anu1217 committed:
Rename element as materials & update references

anu1217 committed:
Change output of make_source

anu1217 committed:
Update SphericalShell/OpenMC_Input/OpenMC_Model/SS_Source_Tallies_Model.py

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Remove volume calculation from make_materials

anu1217 committed:
Calculate volume of depletable material

anu1217 committed:
Update SphericalShell/OpenMC_Input/SS_Model_to_xml.py

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Break main() up into smaller functions

anu1217 committed:
Remove create_openmc_model()

anu1217 committed:
Apply suggestions from code review

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Fix indentation and plot_flux_spectrum input

anu1217 committed:
Break main() function up into smaller functions

anu1217 committed:
Add nested functions inside main()

anu1217 committed:
Update SS_Depletion.py

anu1217 committed:
Update SS_Material.py docstring

anu1217 committed:
Update SS_all_PostProcessing.py

gonuke commented on SphericalShell/OpenMC_Input/SS_Model_to_xml.py:
```diff
@@ -0,0 +1,51 @@
+        tallied_cells = list(spherical_shell_geom.get_all_material_cells().values())
+        talls = make_tallies(tallied_cells)
+        model = openmc.model.Model(geometry = spherical_shell_geom, materials = materials, settings = sets, tallies = talls)
+        return model
```
Functions like this are not normally nested inside the other function.  There are use cases for such nesting, but this probably doesn't make sense here.  They can simply be defined before this main function.

gonuke commented on SphericalShell/OpenMC_Input/SS_Model_to_xml.py:
```diff
@@ -0,0 +1,51 @@
+import argparse
+from OpenMC_SS_Material import alara_element_densities, make_materials
+from OpenMC_SS_Geometry import make_spherical_shell
+from OpenMC_Source_Tallies_Model import make_source, settings, tallies
```
Given your directory structure, I'm not sure these modules will be found.  

This may also be a good time to consider moving these methods into some single file for setting up the common parts of both a steady state and depletion problem.

gonuke commented on SphericalShell/OpenMC_Input/OpenMC_Depletion/SS_Depletion.py:
```diff
@@ -0,0 +1,53 @@
+    chain_file = chain_file_path
+    material.depletable = True
+    volume = 4.0/3.0 * np.pi * ((inner_radius+thickness)**3 - inner_radius**3)
+    material.volume = volume
```
```suggestion
    material.volume = 4.0/3.0 * np.pi * ((inner_radius+thickness)**3 - inner_radius**3)
```

gonuke commented on SphericalShell/OpenMC_Input/OpenMC_Depletion/SS_Depletion.py:
```diff
@@ -0,0 +1,53 @@
+
+def deplete_ss(chain_file_path, model, inner_radius, thickness, times_post_boc, particle_source_rates, norm_mode, timestep_units):
+    material = model.materials[0]
+    chain_file = chain_file_path
```
swapping these lines keeps all the materials modifications together
```suggestion
    chain_file = chain_file_path
    
    material = model.materials[0]
```

gonuke commented on SphericalShell/OpenMC_Input/OpenMC_Depletion/SS_Depletion.py:
```diff
@@ -0,0 +1,53 @@
+    volume = 4.0/3.0 * np.pi * ((inner_radius+thickness)**3 - inner_radius**3)
+    material.volume = volume
+    timesteps = times_post_boc
+    source_rates = particle_source_rates
```
why make additional variables here?

gonuke commented on SphericalShell/OpenMC_Input/OpenMC_Depletion/SS_Depletion.py:
```diff
@@ -0,0 +1,53 @@
+                     dep_params['norm_mode'], 
+                     dep_params['timestep_units'])     
+    
+        integrator.integrate()
```
See earlier comment about nested functions.

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_Depletion_PostProcessing.py:
```diff
@@ -0,0 +1,56 @@
+    nuclide_set = nuclide_set - stable_init_nuc
+    return nuclide_set, materials_object, dep_results, time_steps
+
+def plot_save_data(nuclide_set, materials_object, dep_results, time_steps, nuc_units) :
```
Separation of concerns: This method should perhaps be three difference methods:
* extract summary data into some data structure
* save that data to a text file
* plot that data

gonuke commented on SphericalShell/OpenMC_Output_Processing/SS_all_PostProcessing.py:
```diff
@@ -0,0 +1,52 @@
+                          pp_inputs['indices']['depletable_mat_index'])  
+        data_plot = plot_save_data(nuclide_set, materials_object, dep_results, time_steps, 
+                                   pp_inputs['units']['nuc_units'])
+    
```
See earlier comments about nested functions

gonuke reviewed (changes requested):
Sorry this sat unreviewed for so long.  As the modularity is becoming more clear, it may be time for another refactor.  There are some lingering ideas of modularity by file vs modularity by function.  That is, many of the functions you've created could be in the same file.

I imagine a single file for all your input with functions that handle:
* an argparse lets you choose between depletion or not (as well as other things)
* read YAML
* creating a single spherical shell
* extracting densities from ALARA
* defining materials
* defining tallies
* defining the source
* defining the settings (may have options depending on depletion or not?)
* defining depletion settings
* a function that calls some of these other functions to build an OpenMC model
* a main function that calls the function to build the OpenMC model and either runs the steady state problem and/or runs a depletion problem

For tally processing, we'll aim for a similar single file with various functions to handle all the tasks.

Most (all?) of these functions can simply be cut/paste from the files they are in now with no changes.

anu1217 committed:
Update SphericalShell/OpenMC_Input/OpenMC_Depletion/SS_Depletion.py

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Update SphericalShell/OpenMC_Input/OpenMC_Depletion/SS_Depletion.py

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Separate main() and remove redundant variable names

anu1217 committed:
Separate main and change variable names

anu1217 committed:
Separate main()

anu1217 committed:
Separate plot_save_data into 3 functions

anu1217 committed:
Rename save_data to save_tally_data

anu1217 committed:
Rename functions for depletion and tally data processing

anu1217 committed:
Add Complete_SS_Model.py - includes model & depletion

anu1217 committed:
Update Complete_SS_Model.py

anu1217 committed:
Change variable names

anu1217 committed:
Change variable names

anu1217 committed:
Update SS_all_PostProcessing.py

anu1217 committed:
Remove redundant variable names

anu1217 committed:
Add files via upload

anu1217 commented:
I have added two new files (Complete_SS_Model & Complete_SS_PostProcessing) that does everything needed to build a model, run depletion, and post process. There is now an argparse option that allows the user to choose when depletion is run. I'm guessing the smaller, individual files can now be deleted?

gonuke commented:
Yes - delete the smaller, individual files and we can do a final review of the few files that are left.

anu1217 committed:
Delete SphericalShell/OpenMC_Input/OpenMC_Model/SS_Geometry.py

anu1217 committed:
Delete SphericalShell/OpenMC_Input/OpenMC_Model/SS_Material.py

anu1217 committed:
Delete SphericalShell/OpenMC_Input/OpenMC_Model/SS_Source_Tallies_Model.py

anu1217 committed:
Delete SphericalShell/OpenMC_Input/SS_Model_to_xml.py

anu1217 committed:
Delete SphericalShell/OpenMC_Output_Processing/SS_Depletion_PostProcessing.py

anu1217 committed:
Delete SphericalShell/OpenMC_Output_Processing/SS_Tally_PostProcessing.py

anu1217 committed:
Delete SphericalShell/OpenMC_Output_Processing/SS_all_PostProcessing.py

anu1217 committed:
Delete SphericalShell/OpenMC_Input/OpenMC_Depletion/SS_Depletion.py

gonuke commented:
This looks good for now - I think we've tackled the major issues and have a good basis for moving forward.  There might be room to remove a few layers of subdirectories, but we can tackle that another time
